File: python/29. Divide Two Integers.py
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        a = dividend
        b = divisor
        flag = False
        if (a > 0 and b < 0) or (a < 0 and b > 0):
            flag = True
        a, b = abs(a), abs(b)
        c = b
        mut = 1
        res = 0
        while a > 0 :
            if a >= c:
                res += mut
                a -= c
                c <<= 1
                mut <<= 1
            else:
                c >>= 1
                mut >>= 1
        if flag:
            res = -res
        return min(2147483647, max(-2147483648, res))

File: cpp/1-50/29. Divide Two Integers.cpp
class Solution {
public:
	int divide(long dividend, long divisor) {
		int flag = false;
		if (dividend < INT_MIN || dividend > INT_MAX)
			return INT_MAX;
		if (divisor < INT_MIN || divisor > INT_MAX)
			return INT_MAX;
		if (dividend < 0){
			flag = !flag;
			dividend = -dividend;
		}
		if (divisor < 0){
			flag = !flag;
			divisor = -divisor;
		}
		if(dividend < divisor)
			return 0;
		long res = 0;
		while(dividend > 0 || dividend >= divisor){
			long temp = 1;
			long div = divisor;
			while(div < dividend){
				div <<= 1;
				temp <<= 1;
			}
			if(div > dividend)
			{
				div >>= 1;
				temp>>= 1;
			}
			dividend -= div;
			res += temp;
		}
		res = flag? -res: res;
		if(res < INT_MIN || res > INT_MAX){
			return INT_MAX;
		}
		else
			return res;
	}
};